### PR TITLE
[WIP] Progress on upwindig rules (issue 355)

### PR DESCRIPTION
**Pull request by emmanuellujan**

### CODE REVIEW

**Comment by emmanuellujan** [comment]
So far, cases like the following can be parsed and solved: 

- Dt(u(t,x)) ~ -Dx(u(t,x)) 
- Dt(u(t,x)) ~ Dx(u(t,x))
- Dt(u(t,x)) ~ -c * Dx(u(t,x)) 
- Dt(u(t,x)) ~ -Dx(u(t,x)) + cosh(x)^2 - sinh(x)^2 - 1 

All these cases use Dirichlet BCs.

**Comment by tinosulzer** [comment]
How's this going @emmanuellujan ? Let me know if you need help resolving conflicts.

**Comment by emmanuellujan** [comment]
I have been working on other projects. Next week I can resume this PR, and resolve these conflicts :-)

File: src/MOLFiniteDifference/MOL_discretization.jl [hunk 1]
```diff
@@ -281,16 +284,17 @@ function SciMLBase.symbolic_discretize(pdesys::ModelingToolkit.PDESystem,discret
...
-                # TODO: upwind rules needs interpolation into `@rule`
+
+                # Upwind rules #########################################################
                 forward_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j],II[j]+1]])
                 reverse_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j]-1,II[j]]])
-                # upwinding_rules = [@rule(*(~~a,$(Differential(nottime[j]))(u),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
-                #                         *(~~a..., ~~b..., dot(reverse_weights(II,j),depvars[k][central_neighbor_idxs(II,j)[1:2]])),
-                #                         *(~~a..., ~~b..., dot(forward_weights(II,j),depvars[k][central_neighbor_idxs(II,j)[2:3]]))))
-                #                         for j in 1:nspace, k in 1:length(pdesys.depvars)]
+                upwinding_rules_tmp = [@rule(*(~~a,$(Differential(iv)),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
```
Comment (tinosulzer) [inline src/MOLFiniteDifference/MOL_discretization.jl:294]: The ifelse will work if `a` and `b` can be evaluated. But what if they are dependent variables (e.g. velocity in Navier-Stokes)?

**Review by tinosulzer — COMMENTED** [review]
Looks pretty sensible so far. Note #405 took care of the `IntervalDomain`/`Interval` change

File: src/MOLFiniteDifference/MOL_discretization.jl [hunk 2]
```diff
@@ -289,11 +290,13 @@ function SciMLBase.symbolic_discretize(pdesys::ModelingToolkit.PDESystem,discret
...
                 # Upwind rules #########################################################
-                forward_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j],II[j]+1]])
                 reverse_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j]-1,II[j]]])
-                upwinding_rules_tmp = [@rule(*(~~a,$(Differential(iv)),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
-                                             *(~~a..., ~~b..., dot(reverse_weights(II,j),dv[central_neighbor_idxs(II,j)[1:2]])),
-                                             *(~~a..., ~~b..., dot(forward_weights(II,j),dv[central_neighbor_idxs(II,j)[2:3]]))))
+                forward_weights(II,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, grid[j][[II[j],II[j]+1]])
+                #forward_weights(II,j) = -1.0 * reverse_weights(II,j) # TODO: check this function
+                side = 1.0
+                upwinding_rules_tmp = [@rule(*(~~a,$(Differential(iv))(dv),~~b) => Base.ifelse(*(side, ~~a..., ~~b...,)>0,
```
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference/MOL_discretization.jl:297]: to flip, I think you need `@acrule` @shashi ?

**Comment by emmanuellujan** [comment]
Hi @shashi, I need to create a rule like the following:
```julia
@rule *(~~a, $(Differential(r))(u(x)), ~~b) => "success"
```
where `~~a` and `~~b` are numbers or `nothing`.  Thus, I want to apply the rule here `(-1.0+3.0) * Dx(u(x)) * 2.0` or here `Dx(u(x)) * 2.0`, but not here `-1.0 * Dx(u(x)) * v(x)`
Could you help me? Thank you :)

**Comment by shashi** [comment]
@YingboMa instead of using rule can we use something to get the coefficient of `Differential(r))(u(x)` here?

**Comment by shashi** [comment]
```
    (a, b, islinear) = linear_expansion(t, x)

When `islinear`, return `a` and `b` such that `a * x + b == t`.
```
>but not here `-1.0 * Dx(u(x)) * v(x)`

I don't get how this is different...

**Comment by emmanuellujan** [comment]
> > but not here `-1.0 * Dx(u(x)) * v(x)`
> 
> I don't get how this is different...

 v(x) needs to be evaluated to get a value, but -1.0 is directly a parameter

**Comment by ChrisRackauckas** [comment]
Do you intend to finish this this week or is it just cookie licking now?

**Pull request closed without merging**